pkg/storage/gcp/saver.go

```diff
@@ -0,0 +1,101 @@
+package gcp
+
+import (
+	"fmt"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/bketelsen/buffet"
+	"github.com/gobuffalo/buffalo"
+	"google.golang.org/api/option"
+)
+
+// Storage implements the (github.com/gomods/pkg/storage).Saver interface
+type Storage struct {
+	bucket *storage.BucketHandle
+}
+
+// New returns a new Storage instance
+// authenticated using the provided ClientOptions
+func New(ctx buffalo.Context, cred option.ClientOption) (*Storage, error) {
+	client, err := storage.NewClient(ctx, cred)
+	if err != nil {
+		return nil, fmt.Errorf("could not create new client: %s", err)
+	}
+	// The bucket MUST already exist
+	bkt := client.Bucket("gomodules")
+	return &Storage{bucket: bkt}, nil
+}
+
+// Save uploads the modules .mod, .zip and .info files for a given version
+func (s *Storage) Save(ctx buffalo.Context, module, version string, mod, zip, info []byte) error {
+	sp := buffet.ChildSpan("storage.save", ctx)
+	defer sp.Finish()
+
+	errs := make(chan error, 3)
+	// dispatch go routine for each file to upload
+	// TODO: make this compact
+	go func(errs chan<- error) {
+		sp := buffet.ChildSpan("storage.save.module", ctx)
+		defer sp.Finish()
+
+		if err := save(ctx, s.bucket, fmt.Sprintf("%s/@v/%s.mod", module, version), mod); err != nil {
+			errs <- err
+		} else {
+			errs <- nil
+		}
+	}(errs)
+	go func(errs chan<- error) {
+		sp := buffet.ChildSpan("storage.save.zip", ctx)
+		defer sp.Finish()
+
+		if err := save(ctx, s.bucket, fmt.Sprintf("%s/@v/%s.zip", module, version), zip); err != nil {
+			errs <- err
+		} else {
+			errs <- nil
+		}
+	}(errs)
+	go func(errs chan<- error) {
+		sp := buffet.ChildSpan("storage.save.info", ctx)
+		defer sp.Finish()
+
+		if err := save(ctx, s.bucket, fmt.Sprintf("%s/@v/%s.info", module, version), info); err != nil {
+			errs <- err
+		} else {
+			errs <- nil
+		}
+	}(errs)
+
+	errsOut := make([]string, 3)
+	// wait for each routine above to send a value
+	for count := 0; count < 3; count++ {
+		err := <-errs
```

Comment:
please use some preconfigured timeout when reading from errs. we dont want this to hang

Reply:
like just a check on time elapsed? if so what do you think would be acceptable, 5min? this would be to upload whole zipped modules so I would expect due to variable network speeds it could take a while for larger libraries.

Reply:
something configurable, we can change that any time. 5 minutes is ok to start with. once you hit the timeout log warning.
pattern is like this 
```go
switch{
case err := <-errs: 
    someting
case <- time.After(5 * time.Minutes):
   log.Println("WARNING: upload for moduls M version V, timed out")
}
```

Reply:
or you can apply this pattern on save which makes more sense and gives same timout for each goroutine, benefit: go routine does not hang

```go 
switch {
   case errs <- save(ctx, s.bucket, fmt.Sprintf("%s/@v/%s.mod", module, version), mod):
   case <- time.After(5 * time.Minute): 
      log.Println("WARNING: upload for modul M version V, timed out")
}
```